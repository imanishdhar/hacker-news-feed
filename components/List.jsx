import Link from 'next/link';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faThumbsUp,
  faEyeSlash,
  faComments,
  faUserNinja,
  faHourglassHalf,
  faBlog,
} from '@fortawesome/free-solid-svg-icons';

import { getDomainName, getDifferenceInHrs } from '../utils/commonUtil';

const List = ({
  detail = {},
  index,
  upVoteClickHandler,
  hiddenClickHandler,
}) => (
  <>
    <li key={`${index}-${detail.created_at}`}>
      <section className="flex-container">
        <section className="right-section">
          <div className="icon-container-lg">
            <FontAwesomeIcon icon={faUserNinja} title="author" />
          </div>
          <section>{`@${detail.author}`}</section>
        </section>

        <section>
          <section>
            <span className="title">{detail.title}</span>
          </section>
          <section>
            <section className="inline-block-control">
              <button
                className="icon-container"
                onClick={() => upVoteClickHandler(index)}
              >
                <FontAwesomeIcon
                  icon={faThumbsUp}
                  title="Thumbs up"
                ></FontAwesomeIcon>
              </button>
              <span>{`${detail.points} Points`}</span>
            </section>
            <section className="inline-block-control">
              <button
                className="icon-container"
                onClick={() => hiddenClickHandler(index)}
              >
                <FontAwesomeIcon icon={faEyeSlash} />
              </button>
              <span>{` Hide `}</span>
            </section>
            <section className="inline-block-control">
              <div className="icon-container">
                <FontAwesomeIcon
                  icon={faHourglassHalf}
                  title="hours"
                ></FontAwesomeIcon>
              </div>
              <span>{`${getDifferenceInHrs(
                detail.created_at
              )} Hours Ago`}</span>
            </section>

            <section className="inline-block-control">
              <div className="icon-container">
                <FontAwesomeIcon
                  icon={faComments}
                  title="comments"
                ></FontAwesomeIcon>
              </div>
              <span>{`${detail.num_comments} Comments`}</span>
            </section>
          </section>
          <section>
            {detail.url && (
              <>
                <Link href={detail.url}>
                  <a>
                    <FontAwesomeIcon icon={faBlog} />
                  </a>
                </Link>
                <Link href={detail.url}>
                  <a>{` (${getDomainName(detail.url)}) `}</a>
                </Link>
              </>
            )}
          </section>
        </section>
      </section>

      <style jsx>{`
        button {
          cursor: pointer;
        }

        li {
          font-family: 'Verdana';
          font-size: 13px;
          color: #828282;
          margin: 5px 0;
          border: 1px solid;
          border-radius: 10px;
        }

        a {
          text-decoration: none;
          color: blue;
        }

        a:hover {
          opacity: 0.6;
        }

        section {
          margin: 5px;
        }

        .title {
          color: #000000;
          font-size: 14px;
        }

        .icon-container-lg {
          display: inline-block;
          background: #de7e2b;
          color: #ffffff;
          font-size: 25px;
          padding: 20px;
          border-radius: 50%;
        }

        .flex-container {
          display: block;
          align-items: center;
        }

        .inline-block-control {
          display: block;
        }

        .right-section {
          display: block;
          text-align: center;
          min-width: 120px;
        }

        .icon-container {
          display: inline-block;
          font-size: 11px;
          margin: 0 4px;
        }

        @media only screen and (min-width: 600px) {
          .flex-container {
            display: flex;
          }
          .right-section {
            display: inline-block;
          }
          .icon-container {
            font-size: 15px;
          }
          .inline-block-control {
            display: inline-block;
          }
        }
      `}</style>
    </li>
  </>
);

export default List;
