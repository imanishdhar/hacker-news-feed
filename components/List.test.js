import React from 'react';
import { shallow, render, mount } from 'enzyme';

import List from './List';

function setup() {
  const props = {
    detail: {
      author: 'xyz',
      title: 'Lorem ipsum',
      points: '333',
      num_comments: '777',
      created_at: '2020-04-08T16:56:27.000Z',
      url: 'www.google.com/blog/?page=2&&id=99',
    },
    index: 1,
    key: 'dhjshfghsf',
    upVoteClickHandler: () => {
      mockButtonClicked();
    },
    hiddenClickHandler: () => {
      mockButtonClicked();
    },
  };
  const wrapper = shallow(<List></List>);
  return { wrapper, props };
}

describe('List Test Suite', () => {
  it('Should have a link', () => {
    const { wrapper } = setup();
    expect(wrapper.find('section').exists()).toBe(true);
  });

  it('Should have a button', () => {
    const { wrapper } = setup();
    expect(wrapper.find('button').exists()).toBe(true);
  });

  it('List component should render', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('Upvote button clicked', () => {
    const item = {
      author: 'xyz',
      title: 'Lorem ipsum',
      points: '333',
      num_comments: '777',
      created_at: '2020-04-08T16:56:27.000Z',
      url: 'https://www.google.com',
    };
    const props = {
      index: 1,
      key: 'dhjshfghsf',
    };
    const mockButtonClicked = jest.fn();
    const wrapper = mount(
      <List
        detail={item}
        key={props.key}
        index={props.index}
        upVoteClickHandler={mockButtonClicked}
      ></List>
    );
    const button = wrapper.find('button').at(0);
    button.simulate('click');
    expect(mockButtonClicked).toHaveBeenCalled();
  });
  it('hidden button clicked', () => {
    const item = {
      author: 'xyz',
      title: 'Lorem ipsum',
      points: '333',
      num_comments: '777',
      created_at: '2020-04-08T16:56:27.000Z',
      url: 'https://www.google.com',
    };
    const props = {
      index: 1,
      key: 'dhjshfghsf',
    };
    const mockButtonClicked = jest.fn();
    const wrapper = mount(
      <List
        detail={item}
        key={props.key}
        index={props.index}
        hiddenClickHandler={mockButtonClicked}
      ></List>
    );
    const button = wrapper.find('button').at(1);
    button.simulate('click');
    expect(mockButtonClicked).toHaveBeenCalled();
  });
});
