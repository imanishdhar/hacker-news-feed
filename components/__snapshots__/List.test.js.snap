// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Test Suite List component should render 1`] = `ShallowWrapper {}`;
